sound/soc/sof/sof-audio.c

```diff
@@ -92,14 +129,27 @@ static int sof_widget_free(struct snd_sof_dev *sdev, struct snd_sof_widget *swid
 		.id = swidget->comp_id,
 	};
 	struct sof_ipc_reply reply;
-	int ret;
+	int ret = 0;
 
-	/* only free when refcount is 0 */
 	mutex_lock(&swidget->use_count_mutex);
-	if (--swidget->use_count) {
-		mutex_unlock(&swidget->use_count_mutex);
-		return;
+
+	/* decrement usecount and update config for static DAI widgets */
+	if (swidget->use_count > 1) {
+		if (!swidget->dynamic_pipeline_widget)
+			ret = sof_static_widget_config(sdev, swidget);
+
+		swidget->use_count--;
+		goto unlock;
 	}
+
+	/* retain the static widget if current use_count is 1 */
+	if (swidget->use_count == 1 && !swidget->dynamic_pipeline_widget)
+		goto unlock;
+
+	/* only free when refcount is 0 */
+	if (--swidget->use_count)
+		goto unlock;
+
 	mutex_unlock(&swidget->use_count_mutex);
 
 	if (!swidget->private)
```

Comment:
Hmm, @ranj063 . I also misread the flow a bit.. the free_widget() is a bit misleading as we are not freeing the widget on host, but rather freeing at the DSP side, and use-count  reflect the status on DSP side. Right? 

Still, I think the lock should protect the widget and actions on it, should be atomic. E.g. if while execution this function (related to closing a PCM), another action is done on another CPU that will increase the usage count and that becomes visible to this CPU on L166 here. We'll then proceed to free the host side widget state (including call to "sof_reset_route_setup_status()" and setting complete=0, in this function), while the other CPU is setting up the widget, leaving the widget in an undefined state. This might still be ok if the other CPU takes longer to run, but this doesn't seem correct.

So if we really have risk of concurrent actions on the widget, we should protect the modifications with a lock and keep the lock until the state change is complete (and not just while use_count is changed).

Reply:
if --swidget->use_count > 0, it will return and not change status

Reply:
sorry,  ignore it, I missed something